### PR TITLE
Reject with error when invalid fetch() body

### LINKED ISSUE #2014: fetch() should throw TypeError:  HEAD or GET Request cannot have a body
### What version of Bun is running?

0.5.5

### What platform is your computer?

Darwin 22.2.0 arm64 arm

### What steps can reproduce the bug?

```ts
fetch("https://url", { 
  body: 'content ....' 
})
```

### What is the expected behavior?

Should throw a `TypeError` with some message like: `HEAD or GET Request cannot have a body.`

I am not sure if this is specified in the whatwg standard, but this behavior seems to be standard across browser implementations and nodejs, see also: https://developer.mozilla.org/en-US/docs/Web/API/fetch#exceptions

### What do you see instead?

Instead it seems like a GET or OPTIONS request is executed.

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by ekzhang**
This is a draft PR since I can't figure out how to compile bun on my computer yet! Will work that out and then run tests to make sure it actually builds and runs.

Resolves #2014

### CODE REVIEW

**Comment by ekzhang** [comment]
Update: Okay, just verified it compiles and the test passes!

File: test/bun.js/fetch.test.js [hunk 1]
```diff
@@ -172,6 +172,31 @@ describe("fetch", () => {
...
+    const response = await fetch(url, { method: "POST", body: "buntastic" });
+    expect(response.status).toBe(200);
+    expect(await response.text()).toBe("buntastic");
+
+    // GET cannot have body
+    try {
+      await fetch(url, { body: "buntastic" });
+      expect(false).toBe(true);
+    } catch (exception) {
+      expect(exception instanceof TypeError);
```
Comment (Jarred-Sumner) [inline test/bun.js/fetch.test.js]:
```suggestion
      expect(exception instanceof TypeError).toBe(true);
```

We haven't implemented `expect(exception).toBeInstanceOf(TypeError)` yet

**Comment by Jarred-Sumner** [comment]
Looks good, one nitpicky comment and then I'll merge

**Comment by ekzhang** [comment]
Ah it looks like `exception instanceof TypeError` isn't actually true, and now the test fails.

**Comment by ekzhang** [comment]
Any idea what's happening here? I modified the code in `response.zig` to raise a TypeError using `JSC.toTypeError`, but now when I run:

```js
fetch("https://www.google.com", { body: "hello" }).catch(err => {
  console.log(err);
  console.log(err instanceof TypeError);
});
```

I get this output:

```sh-session
$ packages/debug-bun-darwin-*/bun-debug hello.js
1 | fetch("https://www.google.com", { body: "hello" }).catch((err) => {
   ^
TypeError: fetch() request with GET/HEAD/OPTIONS method cannot have body.
 code: "ERR_INVALID_ARG_VALUE"

      at /Users/ezhang/Documents/Work/bun/hello.js:1:0

false
```

Even though it prints out and displays as a `TypeError`, the instanceof operator still returns false.

image

**Review by jwhear — APPROVED** [review]
Looks good to me, thank you!

**Comment by ekzhang** [comment]
Oh, just FYI the fetch test I added in this PR is still failing @jwhear — not sure if you saw the comment I made above in thread. I was asking for advice on how to resolve that failing test.

**Comment by jwhear** [comment]
Ah, the `instanceof TypeError` issue; I'll look into it in a bit.

**Comment by ThatOneBro** [comment]
I think the problem is that we are creating our own `TypeError` class somewhere which is different from the `TypeError` that we are using globally. I ran into this before in a different area. I'm thinking some of these `TypeError`s  are not like the others.

**Pull request merged**